### ISSUE TITLE
Syntax highlighting

### ISSUE THREAD

**Issue by gitonthescene** [exercism/bqn#5]
It’s probably worth having separate issues for subtasks of the main issue.  I’ll transfer info from the main issue here.

**Comment by gitonthescene** [comment]
- @ErikSchierboom Outlined the basic ask [here](https://github.com/exercism/bqn/issues/1#issuecomment-1269493872)
  - @razetime pointed at a [codemirror plugin](https://github.com/exercism/bqn/issues/1#issuecomment-1269454911) we might be able to use wholesale.

**Comment by gitonthescene** [comment]
@ErikSchierboom - Is there any [doc you can link](https://github.com/exercism/bqn/issues/1#issuecomment-1269493872) to to show exactly where these hooks are?

Searches under exercism/ for “syntax” and “highlighting” turned up nothing.  Search for “codemirror” turned just this [wren related project](https://github.com/exercism/codemirror-lang-wren).  A search for “editor” resulted only in that same link.

**Comment by ErikSchierboom** [comment]
@gitonthescene The highlightjs configuration is in the `config.json` file. It is mentioned as part of the ["Prepare for launch" document](https://exercism.org/docs/building/tracks/new/prepare-for-launch) which is linked to in https://github.com/exercism/bqn/issues/1. 

The codemirror plugin loading is done here: https://github.com/exercism/website/blob/main/app/javascript/components/misc/CodeMirror/languageCompartment.ts

**Comment by gitonthescene** [comment]
@ErikSchierboom - Awesome!  Thanks very much.

**Comment by gitonthescene** [comment]
@ErikSchierboom - FWIW, it looks like searching at the org level defaults to looking for repository names.  I needed to switch to the Code tab to find the info you linked.  Thanks again.

**Comment by ErikSchierboom** [comment]
> Am I right to read this as meaning the language must be registered at [highlight.js](https://github.com/highlightjs/highlight.js/blob/main/SUPPORTED_LANGUAGES.md) since it’s a required field?

No, it could also be a highlightjs plugin, it doesn't have to be registered anywhere (well, except for NPM).

**Comment by andreypopp** [comment]
just FYI, it's possible to reuse codemirror highlighting to highlight static text (outside of codemirror editor), [example code here](https://github.com/mechanize-systems/bqnpad/blob/main/lang-bqn/index.ts#L825-L842).

**Comment by gitonthescene** [comment]
@andreypopp - This sounds like how Exercism itself should be set up rather than requiring two different methods.

@ErikSchierboom

**Comment by ErikSchierboom** [comment]
What do you mean?

**Comment by gitonthescene** [comment]
@ErikSchierboom - I’m not sure where the confusion is.  Instead of using highlight.js in the linked file we could just use the codemirror plugin as suggested in the code that @andreypopp linked above.

Is there any reason not to use the codemirror plugin for stand-alone highlighting?

Obviously allowing this doesn’t preclude using highlight.js when it has ben configured.  But reusing the codemirror highlighting ensures consistency in addition to being less work.

**Comment by gitonthescene** [comment]
@iHiD Is it really that much overhead over loading highlight.js?  Again, using the same component ensures consistency.

**Comment by iHiD** [comment]
The overhead would be loading codemirror. But yes, turning every place that solutions are rendered on the site into JavaScript components rather than just HTML would be a big increase. Consider that we have pages everywhere of dozens of solutions: https://exercism.org/profiles/ErikSchierboom/solutions - all of the rendering for those would move client-side, and thus would be slower, would be worse for SEO, etc.

**Comment by andreypopp** [comment]
> all of the rendering for those would move client-side, and thus would be slower

Not really, the snippet of code I've linked above can be run on server at the time of static site generation, you'd need to change callback to produce HTML strings instead of build DOM nodes. That way there won't be any codemirror code loaded at runtime.

**Comment by ErikSchierboom** [comment]
> Not really, the snippet of code I've linked above can be run on server at the time of static site generation

Yes, but we're running Ruby, not JavaScript on our backend. We could shell out to Nodejs to run code, but that would _also_ mean getting a performance hit.

While we appreciate your thoughts on this, we won't be replacing highlightjs with codemirror. Highlightjs has worked great for us, is more performant than codemirror _and_ supports more languages.

**Comment by andreypopp** [comment]
not really pushing on anything, just tried to provide factually correct  information about what codemirror highlighting infra is :) sorry for the noise

**Comment by ErikSchierboom** [comment]
No problem. Thanks.

**Comment by gitonthescene** [comment]
@ErikSchierboom So it looks like we have a [published codemirror plugin](https://github.com/mechanize-systems/bqnpad/issues/7#issuecomment-1279982990) to register.  Is there anything we should do to test the plugin once it's been [added](https://github.com/exercism/website/blob/main/app/javascript/components/misc/CodeMirror/languageCompartment.ts).

**Comment by ErikSchierboom** [comment]
We'll add the track to the website, but still inactive. That way you'll be able to test the track without any students being able to access the track. I'll let you know when I've had time to do so.

**Comment by gitonthescene** [comment]
Super.  Thanks very much.

**Comment by razetime** [comment]
started a [highlightjs plugin](https://github.com/razetime/highlightjs-bqn). React with a thumbs up to this message if you'd like contributor access.

**Comment by ErikSchierboom** [comment]
@gitonthescene See https://github.com/exercism/website/pull/3165

**Comment by gitonthescene** [comment]
@ErikSchierboom Thanks for doing this.  I’m traveling at the moment but will have a look at what’s up when I’m back.

**Comment by razetime** [comment]
so. both codemirror and a highlightjs plugin are complete. Do I have to get an npm package in?

**Comment by razetime** [comment]
have published it at https://www.npmjs.com/package/highlightjs-bqn. Hopefully it works correctly with the highlightjs node plugin.

**Comment by ErikSchierboom** [comment]
@razetime Seems to be working. I've opened a PR: https://github.com/exercism/website/pull/3397
You'll need to update the `config.json` to change the highlightjs language to bqn.

**Comment by ErikSchierboom** [comment]
PR has been merged!
